### PR TITLE
New serving

### PR DESCRIPTION
**Pull request by jplu**
# What does this PR do?

This PR proposes a new way to create a saved model to be properly served via TF Serving. The logic behind is to create a `serving` that will be used to create the expected saved model with a proper input signature. Currently the saved models are very limited:

- input sequence length limited to exactly 5 tokens
- input parameters limited to have only `input_ids`
- When `output_attentions` or `output_hidden_states` was set to True, the saved model output contained as many outputs as the number of attentions or hidden state

This PR fixes these 3 issues. A new behavior is also introduced, when doing `model.save_pretrained(...)` a saved model version is also created in same time than the `.h5` weights file.

The proposed logic allows anybody to create its own input signature simply by overwriting the new `serving` method. For example, the default inputs for BERT are now `input_ids`, `attention_mask` and `token_type_ids`, if one wants to replace `input_ids` by `inputs_embeds`, a new model has to be created overwriting the `serving` method like:

```
class CustomBertModel(TFBertModel):
    @tf.function(
        input_signature=[
            {
                "inputs_embeds": tf.TensorSpec((None, None, 768), tf.float32, name="inputs_embeds"),
                "attention_mask": tf.TensorSpec((None, None), tf.int32, name="attention_mask"),
                "token_type_ids": tf.TensorSpec((None, None), tf.int32, name="token_type_ids"),
            }
        ]
    )
    def serving(self, inputs):
        output = self.call(inputs)

        return self.serving_output(output)

model = CustomBertModel.from_pretrained("bert-base-cased")
model.save_pretrained("saving_path")
```
Slow/quick tests are passing.

EDIT: ping @sgugger @patrickvonplaten and @LysandreJik for review.

### CODE REVIEW

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:569]: some leftover comments here I think

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:616]: Maybe add a comment that each model should implement this function here

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:592]: is `token_type_ids` general enough to be here? I'm fine with `attention_mask`

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:853]: What is a version of a model? This looks dangerous

Comment (patrickvonplaten) [reply src/transformers/modeling_tf_utils.py:853]: Do we need that?

File: tests/test_modeling_tf_common.py
Comment (patrickvonplaten) [inline tests/test_modeling_tf_common.py:167]: Isn't that tested already that model can be saved and reloaded in TF (the normal way)

Comment (patrickvonplaten) [reply tests/test_modeling_tf_common.py:167]: Ah ok, I see...I think we should maybe switch the default of `saved_model` to `False` to prevent backwards breaking changes.

File: src/transformers/modeling_tf_utils.py
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:842]: The default should be `False` here IMO to not have backwards breaking changes without notifying the user. We can switch the flip in a future version. Also should this `saved_model` param maybe go in the config under a different name like `save_in_optimized_format` ? and we let the model default to the config?

File: src/transformers/modeling_tf_utils.py
Comment (jplu) [reply src/transformers/modeling_tf_utils.py:592]: As far as I see, `token_type_ids` appears in more models than not.

File: src/transformers/modeling_tf_utils.py
Comment (jplu) [reply src/transformers/modeling_tf_utils.py:853]: One of the advantages of the saved models is that they can be versioned. The version is very important as it is required to be loaded by TF serving, otherwise the server raise an error.

**Review by patrickvonplaten — COMMENTED** [review]
In general the PR looks good to me! 

Two things I think we should change:

1) The default for `saved_model` should be `False` (at least for now) IMO. Otherwise all model saving will take significant longer than expected -> big breaking change that's easy to prevent IMO.

2) I think it makes sense to add something like `save_tf_in_optimized_format` to the config and let the save function default to that.

File: src/transformers/modeling_tf_utils.py
Comment (jplu) [reply src/transformers/modeling_tf_utils.py:842]: I don't mind having this parameter to false by default^^

File: src/transformers/modeling_tf_utils.py
Comment (jplu) [reply src/transformers/modeling_tf_utils.py:592]: My logic behind was to take the same inputs their those given by their respective tokenizer

Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py:592]: If you follow the tokenizer logic (as I think you should!), then you're right to define both attention mask and token type ids here, as that's [what's done in `model_input_names`](https://github.com/huggingface/transformers/blob/master/src/transformers/tokenization_utils_base.py#L1531).

Since you're overriding that in models down the road, this change LGTM.

File: src/transformers/modeling_tf_utils.py
Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py:842]: Yes, I would put it to `False` by default as well.

File: src/transformers/modeling_tf_utils.py
Comment (LysandreJik) [reply src/transformers/modeling_tf_utils.py:853]: If it's a necessary parameter required by TF serving then ok for me.

File: src/transformers/modeling_tf_utils.py
Comment (LysandreJik) [inline src/transformers/modeling_tf_utils.py:863]: Does TF serving need that exact path? i.e., `dir/saved_model/{version}`?

File: src/transformers/models/distilbert/modeling_tf_distilbert.py
Comment (LysandreJik) [inline src/transformers/models/distilbert/modeling_tf_distilbert.py:494]: Since this gets overridden in models that have no `token_type_ids`, I find this very clean!

File: src/transformers/models/albert/modeling_tf_albert.py
Comment (LysandreJik) [inline src/transformers/models/albert/modeling_tf_albert.py:1286]: I'm assuming all losses are set to `None` because no training happen in `TFServing`, right?

**Review by LysandreJik — APPROVED** [review]
This looks great! I see no backwards-compatibility issues once the saving flag has been changed to `False` by default.

This looks like the cleanest solution by far, good job!

File: src/transformers/modeling_tf_utils.py
Comment (jplu) [reply src/transformers/modeling_tf_utils.py:863]: Yes, it does, you can take a look at this basic serving example https://www.tensorflow.org/tfx/serving/serving_basic the exported model has a version number to be able to be used.

File: src/transformers/models/albert/modeling_tf_albert.py
Comment (jplu) [reply src/transformers/models/albert/modeling_tf_albert.py:1286]: Yes, no training through TFServing, then loss to `None` to make it disappear from the future output.

**Review by patrickvonplaten — APPROVED** [review]
Cool, good to merge for me as well then!

**Comment by jplu** [comment]
Ok, the test `test_saved_model_creation` is skipped if it needs more than 30sec to be executed for a model. For now these models are skipped:
BART
BlenderBot
Funnel
Longformer
Lxmert
Marian
MBart
Mobilebert
Pegasus
T5

Let's see if the test becomes faster once I will optimise these model like I did for BERT. LGTM!

**Comment by patrickvonplaten** [comment]
Cool, maybe @sgugger can take a look as well :-)

**Comment by jplu** [comment]
Oops forgot to rebase and the changes for the LED model is missing, and also the changes in the Seq2Seq template. Please wait my next push before merging.

File: src/transformers/modeling_tf_utils.py
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:850]:
```suggestion
            saved_model (:obj:`bool`, `optional`, defaults to :obj:`False`):
```

File: src/transformers/modeling_tf_utils.py
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:852]:
```suggestion
            version (:obj:`int`, `optional`, defaults to 1):
```

File: src/transformers/modeling_tf_utils.py
Comment (sgugger) [reply src/transformers/modeling_tf_utils.py:853]: Could we expand the description and link to the relevant doc page on the TF serving side?

File: src/transformers/models/albert/modeling_tf_albert.py
Comment (sgugger) [reply src/transformers/models/albert/modeling_tf_albert.py:1286]: Since it's the default in all `ModelOutput`, why not omit it entirely?

File: src/transformers/models/bart/modeling_tf_bart.py
Comment (sgugger) [inline src/transformers/models/bart/modeling_tf_bart.py:1136]: The formatter output is pretty awful there. Could we put those in intermediate variables just before to try and get something more readable?

File: src/transformers/models/bart/modeling_tf_bart.py
Comment (sgugger) [inline src/transformers/models/bart/modeling_tf_bart.py:1302]: Same here.

File: src/transformers/models/dpr/modeling_tf_dpr.py
Comment (sgugger) [inline src/transformers/models/dpr/modeling_tf_dpr.py:746]: Is it a previous mistake fixed here? It doesn't seem relevant to serving.

File: src/transformers/models/lxmert/modeling_tf_lxmert.py
Comment (sgugger) [inline src/transformers/models/lxmert/modeling_tf_lxmert.py:1053]: Same as with BART here.

File: src/transformers/models/lxmert/modeling_tf_lxmert.py
Comment (sgugger) [inline src/transformers/models/lxmert/modeling_tf_lxmert.py:1494]: Same here.

File: src/transformers/models/t5/modeling_tf_t5.py
Comment (sgugger) [inline src/transformers/models/t5/modeling_tf_t5.py:1199]: Same as with BART here.

File: src/transformers/models/t5/modeling_tf_t5.py
Comment (sgugger) [inline src/transformers/models/t5/modeling_tf_t5.py:1426]: And again here

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py
Comment (sgugger) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py:898]: Those should fit on one line.

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py
Comment (sgugger) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py:1008]: Same here.

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py
Comment (sgugger) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py:1131]: Here too.

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py
Comment (sgugger) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py:1287]: Here too.

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py
Comment (sgugger) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py:1387]: And here

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py
Comment (sgugger) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py:2467]: For those, we should do like BART and have the long lines as intermediate results before (on one line if possible).

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py
Comment (sgugger) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_tf_{{cookiecutter.lowercase_modelname}}.py:2632]: Same here.

**Review by sgugger — APPROVED** [review]
Thanks for adding this! It looks good to me (with the comments from Lysanbdre and Patrick). Just two remarks:
- the formatter is outputing something pretty awful for the BART (and the likes) serving methods, there is probably some way to have the code clearer
- there is a lot of dupe code accross files, so I think we should take advantage of the current scripts and add some `# Copied from xxx` to make sure all the copies stay in sync and are easily updated.

File: src/transformers/models/dpr/modeling_tf_dpr.py
Comment (jplu) [reply src/transformers/models/dpr/modeling_tf_dpr.py:746]: It is a small fix yes!

**Comment by jplu** [comment]
I should have addressed all the comments. The saved model creation tests are silent for the Seq2Seq models until I find a proper fix.

**Comment by patrickvonplaten** [comment]
Great should we merge @jplu  @LysandreJik @? - it's blocking the TF-Bart Split PR a bit.

**Comment by jplu** [comment]
For me it is good to merge if there is no other comments :)

**Comment by patrickvonplaten** [comment]
Cool merging then

**Comment by sgugger** [comment]
I had some more comments actually! With the short names, most of the outputs of the serving methods fit on one line now. black does not put things back on the same line once it has split on several, so it's not fixed by the quality scripts.

I also think it would make future maintenance easier to add the # Copied from comments for dupe code.

**Comment by jplu** [comment]
They are all on one line (when it is possible, which means not too many characters to fit in).

I will open a PR to take care of adding the `#copied from` comments once I finish to fix the S2S models.

File: src/transformers/models/albert/modeling_tf_albert.py
Comment (sgugger) [inline src/transformers/models/albert/modeling_tf_albert.py:1091]: This fits in less than 119 characters with
```
        return TFMaskedLMOutput(logits=output.logits, hidden_states=hs, attentions=attns)
```

**Comment by sgugger** [comment]
See comment above, and this is just one example, most of those now fit in one line with your last changes.

File: src/transformers/models/albert/modeling_tf_albert.py
Comment (jplu) [reply src/transformers/models/albert/modeling_tf_albert.py:1091]: AH those ones! Sorry I thought you were talking about only the `if self.config.use_cache...` things. Ok nevermind then, I will also take care of this!

**Pull request merged**
